lib/ruby/truffle/jruby+truffle/lib/runner.rb

```diff
@@ -400,73 +444,144 @@ def subcommand_run(rest)
+  def subcommand_readme(rest)
+    puts File.read(ROOT + 'README.md')
+  end
 
-    puts "executing #{commands.size} commands:"
-    commands.each { |cmd| print_cmd cmd, true }
-    puts
+  def subcommand_ci(rest)
+    if (path = options[:ci][:batch])
+      batch = if path =~ /^in|stdin$/
+                $stdin.read
+              else
+                File.read(path)
+              end
+
+      results = batch.each_line.map do |line|
+        next if line =~ /^#/ || line.strip.empty?
 
-    commands.each do |cmd|
-      unless execute_cmd(cmd, fail: false, print_always: true)
-        exit $?.exitstatus
+        options       = {}
+        option_parser = build_option_parser OPTION_DEFINITIONS[:ci], options
+        rest          = option_parser.order line.split
+
+        gem_name = rest.first
+        CIEnvironment.new(options, self.options[:global][:dir], gem_name).success?
       end
+
+      exit results.all? ? 0 : 1
+    else
+      gem_name = rest.first
+      CIEnvironment.new options[:ci], options[:global][:dir], gem_name
     end
   end
 
-  def get_stored_command(name, fail: true)
-    result = @options.fetch(:stored_commands, {})[name.to_sym]
-    raise("unknown stored command: #{name}") if fail && !result
-    Array(result).flat_map { |c| (c.is_a? Symbol) ? get_stored_command(c) : c }
+  def execute_cmd(cmd, dir: nil, raise: true, print_always: false)
+    super cmd, dir: dir, raise: raise, print: verbose? || print_always
   end
 
-  def subcommand_clean(rest)
-    FileUtils.rm_rf @options[:global][:truffle_bundle_path]
-  end
+  class CIEnvironment
+    include Utils
 
-  def subcommand_readme(rest)
-    readme_path = File.join File.dirname(__FILE__), '../README.md'
-    puts File.read(readme_path)
-  end
+    def self.define_dsl_attr(*names, &conversion)
+      nothing = Object.new
 
-  def print_cmd(cmd, print_always)
-    unless verbose? || print_always
-      return cmd
+      names.each do |name|
+        ivar_name = :"@#{name}"
+
+        define_method name do |value = nothing|
+          if value == nothing
+            instance_variable_get ivar_name
+          else
+            conversion ||= -> v { v }
+            instance_variable_set ivar_name, conversion.call(value)
+          end
+        end
+      end
     end
 
-    print = if String === cmd
-              cmd
-            else
-              cmd.map do |v|
-                if Hash === v
-                  v.map { |k, v| "#{k}=#{v}" }.join(' ')
-                else
-                  Shellwords.escape v
-                end
-              end.join(' ')
-            end
-    puts '$ ' + print
-    return cmd
-  end
+    define_dsl_attr :repository_name, :subdir
+    define_dsl_attr(:working_dir) { |v| Pathname(v) }
+    attr_reader :gem_name
+
+    def initialize(options, working_dir, gem_name)
+      @options  = options
+      @gem_name = gem_name
+
+      ci_file = Dir.glob(ROOT + 'gem_ci' + "#{@gem_name}.rb").first ||
```

Comment:
I find `Pathname#/` clearer but it's Ruby 2.2+. Maybe `Pathname#join` in this particular case?
Also the Dir.glob seems fancy for `[path1, path2].find(&:exist?)` if I read right.